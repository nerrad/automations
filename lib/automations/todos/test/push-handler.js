const {
	gimmeOctokit,
	gimmeContext,
	gimmeApp,
	loadDiff,
} = require( '@testHelpers' );
const { reset } = require( '../utils/check-for-duplicate-issue' );

describe( 'push-handler', () => {
	let octokit, context, app;

	beforeEach( () => {
		octokit = gimmeOctokit();
		context = gimmeContext( 'push' );
		app = gimmeApp( 'todos' );
		reset();
	} );

	it( 'creates an issue', async () => {
		await app( context, octokit );
		expect( octokit.issues.create ).toHaveBeenCalledTimes( 1 );
		expect( octokit.issues.create.mock.calls[ 0 ] ).toMatchSnapshot();
	} );
	it( 'creates an issue with a @todo comment', async () => {
		octokit.repos.getCommit.mockReturnValue( loadDiff( 'at-todo' ) );
		await app( context, octokit );
		expect( octokit.issues.create ).toHaveBeenCalledTimes( 1 );
		expect( octokit.issues.create.mock.calls[ 0 ] ).toMatchSnapshot();
	} );
	it( 'creates an issue with a truncated title', async () => {
		octokit.repos.getCommit.mockReturnValue( loadDiff( 'long-title' ) );
		await app( context, octokit );
		expect( octokit.issues.create ).toHaveBeenCalledTimes( 1 );
		expect( octokit.issues.create.mock.calls[ 0 ] ).toMatchSnapshot();
	} );
	it( 'does not create any issues if no todos are found', async () => {
		octokit.repos.getCommit.mockReturnValue( loadDiff( 'none' ) );
		await app( context, octokit );
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
	it( 'does not create any issues if the push is not on the default branch', async () => {
		context.payload = { ...context.payload, ref: 'not-master' };
		await app( context, octokit );
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
	it( 'does not create an issue that already exists', async () => {
		octokit.search.issuesAndPullRequests.mockReturnValueOnce(
			Promise.resolve( {
				data: {
					total_count: 1,
					items: [
						{ title: 'I am an example title', state: 'open' },
					],
				},
			} )
		);
		await app( context, octokit );
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
	it( 'does not create the same issue twice in the same run', async () => {
		octokit.repos.getCommit.mockReturnValue( loadDiff( 'duplicate' ) );
		await app( context, octokit );
		expect( octokit.issues.create ).toHaveBeenCalledTimes( 1 );
	} );
	it( 'creates an issue if the search does not have an issue with the correct title', async () => {
		octokit.search.issuesAndPullRequests.mockReturnValueOnce(
			Promise.resolve( {
				data: {
					total_count: 1,
					items: [ { title: 'Not found', state: 'open' } ],
				},
			} )
		);
		await app( context, octokit );
		expect( octokit.issues.create ).toHaveBeenCalledTimes( 1 );
	} );
	it( 'creates many (5) issues', async () => {
		octokit.repos.getCommit.mockReturnValue( loadDiff( 'many' ) );
		await app( context, octokit );
		expect( octokit.issues.create ).toHaveBeenCalledTimes( 5 );
		expect( octokit.issues.create.mock.calls ).toMatchSnapshot();
	} );
	it( 'ignores pushes not to master', async () => {
		context.payload = { ...context.payload, ref: 'not/the/master/branch' };
		await app( context, octokit );
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
	it( 'ignores merge commits', async () => {
		octokit.git.getCommit.mockReturnValueOnce(
			Promise.resolve( {
				data: { parents: [ 1, 2 ] },
			} )
		);
		await app( context, octokit );
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
	it( 'reopens a closed issue', async () => {
		octokit.search.issuesAndPullRequests.mockReturnValueOnce(
			Promise.resolve( {
				data: {
					total_count: 1,
					items: [
						{
							number: 1,
							title: 'I am an example title',
							state: 'closed',
						},
					],
				},
			} )
		);
		await app( context, octokit );
		expect( octokit.issues.update ).toHaveBeenCalledTimes( 1 );
		expect( octokit.issues.createComment ).toHaveBeenCalledTimes( 1 );
		expect(
			octokit.issues.createComment.mock.calls[ 0 ]
		).toMatchSnapshot();
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
	it( 'does nothing with a diff over the max size', async () => {
		octokit.repos.getCommit.mockReturnValue(
			Promise.resolve( { headers: { 'content-length': 2000001 } } )
		);
		await app( context, octokit );
		expect( octokit.issues.create ).not.toHaveBeenCalled();
	} );
} );
