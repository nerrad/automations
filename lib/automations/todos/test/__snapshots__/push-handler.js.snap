// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`push-handler creates an issue 1`] = `
Array [
  Object {
    "body": "I am an example title

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/.travis.yml#L9-L15
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
    "owner": "nerrad",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;

exports[`push-handler creates an issue with a @todo comment 1`] = `
Array [
  Object {
    "body": "I am an example title

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/.travis.yml#L9-L15
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
    "owner": "nerrad",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;

exports[`push-handler creates an issue with a truncated title 1`] = `
Array [
  Object {
    "body": "This is a really, really, really, really, really, really, really, really, really, really long title

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/.travis.yml#L9-L15
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
    "owner": "nerrad",
    "repo": "tests",
    "title": "This is a really, really, really, really, really, really,...",
  },
]
`;

exports[`push-handler creates many (5) issues 1`] = `
Array [
  Array [
    Object {
      "body": "Title number 1 multiline

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/file.txt#L4-L5
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
      "owner": "nerrad",
      "repo": "tests",
      "title": "Title number 1 multiline",
    },
  ],
  Array [
    Object {
      "body": "Title number 2

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/file.txt#L4
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
      "owner": "nerrad",
      "repo": "tests",
      "title": "Title number 2",
    },
  ],
  Array [
    Object {
      "body": "Title number 3

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/file.txt#L4
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
      "owner": "nerrad",
      "repo": "tests",
      "title": "Title number 3",
    },
  ],
  Array [
    Object {
      "body": "Title number 4 This is some other sentence with todo in it.

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/file.txt#L4-L6
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
      "owner": "nerrad",
      "repo": "tests",
      "title": "Title number 4 This is some other sentence with todo in it.",
    },
  ],
  Array [
    Object {
      "body": "Title number 5

---
https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/file.txt#L4
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in e06c237a0c041f5a0a61f1c361f7a1d6f3d669af. cc @nerrad",
      "owner": "nerrad",
      "repo": "tests",
      "title": "Title number 5",
    },
  ],
]
`;

exports[`push-handler reopens a closed issue 1`] = `
Array [
  Object {
    "body": ":rocket: This issue has been reopened because the **\`todo\`** comment still exists in [****](https://github.com/nerrad/tests/blob/e06c237a0c041f5a0a61f1c361f7a1d6f3d669af/), as of e06c237a0c041f5a0a61f1c361f7a1d6f3d669af.

---
###### If this was not intentional, just remove the comment from your code.",
    "issue_number": 1,
    "owner": "nerrad",
    "repo": "tests",
  },
]
`;
